src/react/hooks.js

```diff
@@ -0,0 +1,86 @@
+const { useContext, useEffect, useState, useRef } = require('react')
+const { AtomContext, differ } = require('../react')
+const raf = require('../raf')
+
+const isServer = typeof navigator === 'undefined'
+
+function useAtom (selector, options = {}) {
+  selector = selector || (state => state)
+  const { sync = false, pure = true, observe = !isServer } = options
+  const schedule = sync ? fn => () => fn() : raf
+  const { atom } = useContext(AtomContext)
+  assert(atom, 'No atom found in context, did you forget to wrap your app in <Provider atom={atom} />?')
+  const [mappedProps, setMappedProps] = useState(selector(atom.get()))
+  const ref = useRef({ mappedProps, atom })
+
+  // in case atom is swapped, unobserve
+  if (ref.current.unobserve && atom !== ref.current.atom) {
+    ref.current.unobserve()
+    ref.current.unobserve = null
+  }
+
+  if (!ref.current.unobserve && observe) {
+    ref.current.unobserve = atom.observe(() => {
+      // store updates happening in rapid sequence
+      // get cancelled and rescheduled
+      if (ref.current.cancelUpdate) {
+        ref.current.cancelUpdate()
+        ref.current.cancelUpdate = null
+      }
+      // schedule an update
+      ref.current.cancelUpdate = schedule(() => {
+        ref.current.cancelUpdate = null
+        const nextMappedProps = selector(atom.get())
+        if (!pure || differ(ref.current.mappedProps, nextMappedProps)) {
+          setMappedProps(nextMappedProps)
+        }
+      })()
+    })
+  }
+
+  // cancel a scheduled update on render
+  // we just rendered - don't need to render again
+  if (ref.current.cancelUpdate) {
+    ref.current.cancelUpdate()
+    ref.current.cancelUpdate = null
+  }
+
+  // unsubscribe from the store
+  useEffect(
+    () => {
+      // note, we already subscribed on first render
+      // this effect is only for unsubscribing
```

Comment:
why aren't we subscribing inside this effect? You are effectively subscribing in a render fn right now

Reply:
Yeah, so.. this does seem a bit dodgy. I'm not sure if this is gonna fly long term, like, I imagine there could be bugs around this logic of subscribing inline and not in effect when it comes to async rendering etc.

But I'm actually also doing something similar in the old connectors, I subscribe to atom in render.

This is because I'm trying to control the order of subscriptions, so that we always trigger rerenders in the direction of parent -> child. If you don't do it in that order, things break, e.g. if you remove some state that child depends on (say if child is a modal), but you don't rerender parent first, which is meant to remove the modal, the child (modal) rerenders first and throws an exception about missing state, etc.

So really, to ensure correct order of rerendering and minimum number of renders, I couldn't figure out how I'd use useEffect for that (or componentDidMount for that matter). Because:

<img width="535" alt="image" src="https://user-images.githubusercontent.com/324440/49434716-b51fdd00-f7ac-11e8-9d9c-d0f372a70090.png">

And scenarios that this takes into account now are:

1. State changes in a way that affects all 3 components. Component A rerenders once, also rerendering it's children, which extract up to date latest state. We cancel store scheduled rerenders of Components B and C. 👍
2. State changes in a way that only affects Component A and component C and component B is wrapped in React.memo, it's pure. So Component A rerenders first, trying to rerender component B, but that one's pure, so nothing happens. Then Component C is rerendered, because it's also observing the store and the update wasn't cancelled this time, so it also correctly rerenders. 👍
3. All components are wrapped in React.memo. It's fine, they all individually rerender. 👍
4. State changes, Component B diffs mapped props, nothing changed, doesn't call setState doesn't rerender. Component C does the same and mapped props changed, it rerenders. 👍
5. Component B renders a new subtree, like a modal with nested components inside of it, Component B1, Component B2. Everything still works, because subscriptions fire in order, so all of the above examples still hold. 👍

And that's why I subscribe inline, resubscribe inline and only use useEffect for disposing of subscription. I would like to use useEffect, but just not sure how to keep all of those 5 behaviours above.

I used to think you just need to fire atom observers in reverse order. But that's not true for things that are rendered later in time, like modal. Because that's when we noticed a bug in User UI back in the day, where modal was trying to rerender, but the state has already been cleared, but the parent hasn't rerendered yet, it was wrong order. It's not just about direction, it's this whole tree and the only way I found a way to keep track of order was by using renders.

Maybe, it's possible to only keep a counter in render(), like atom.order()=> keeps returning a globally incrementing counter, now you know the order of renders. And you only subscribe in useEffect, like you're meant to, but you use the order number to put the subscribed in the right slot. Need to think about it.

Reply:
I changed this, I think it works, I keep track of rendering order in a ref and then subscribe to things in that order. This way I get the benefits I was looking for, but I keep the effectful logic in useEffect!